gobblin-utility/src/main/java/gobblin/util/logs/LogCopier.java

```diff
@@ -0,0 +1,446 @@
+import com.google.common.util.concurrent.AbstractScheduledService;
+
+import gobblin.util.ExecutorsUtils;
+import gobblin.util.FileListUtils;
+
+
+/**
+ * A utility class that periodically reads log files in a source log file directory for changes
+ * since the last reads and appends the changes to destination log files with the same names as
+ * the source log files in a destination log directory. The source and destination log files
+ * can be on different {@link FileSystem}s.
+ *
+ * <p>
+ *   This class extends the {@link AbstractScheduledService} so it can be used with a
+ *   {@link com.google.common.util.concurrent.ServiceManager} that manages the lifecycle of
+ *   a {@link LogCopier}.
+ * </p>
+ *
+ * <p>
+ *   This class is intended to be used in the following pattern:
+ *
+ *   <pre>
+ *     {@code
+ *       LogCopier.Builder logCopierBuilder = LogCopier.newBuilder();
+ *       LogCopier logCopier = logCopierBuilder
+ *           .useSrcFileSystem(FileSystem.getLocal(new Configuration()))
+ *           .useDestFileSystem(FileSystem.get(URI.create(destFsUri), new Configuration()))
+ *           .readFrom(new Path(srcLogDir))
+ *           .writeTo(new Path(destLogDir))
+ *           .useInitialDelay(60)
+ *           .useCopyInterval(60)
+ *           .useTimeUnit(TimeUnit.SECONDS)
+ *           .build();
+ *
+ *       ServiceManager serviceManager = new ServiceManager(Lists.newArrayList(logCopier));
+ *       serviceManager.startAsync();
+ *
+ *       // ...
+ *       serviceManager.stopAsync().awaitStopped(60, TimeUnit.SECONDS);
+ *     }
+ *   </pre>
+ *
+ *   Checkout the Javadoc of {@link LogCopier.Builder} to see the available options for customization.
+ * </p>
+ *
+ * @author ynli
+ */
+public class LogCopier extends AbstractScheduledService {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(LogCopier.class);
+
+  private static final long DEFAULT_SOURCE_LOG_FILE_MONITOR_INTERVAL = 120;
+  private static final long DEFAULT_LOG_COPY_INTERVAL_SECONDS = 60;
+  private static final TimeUnit DEFAULT_TIME_UNIT = TimeUnit.SECONDS;
+  private static final int DEFAULT_LINES_WRITTEN_BEFORE_FLUSH = 100;
+
+  private final FileSystem srcFs;
+  private final FileSystem destFs;
+  private final Path srcLogDir;
+  private final Path destLogDir;
+
+  private final long sourceLogFileMonitorInterval;
+  private final long copyInterval;
+  private final TimeUnit timeUnit;
+
+  private final Set<String> logFileExtensions;
+
+  private final Optional<Pattern> filterRegexPattern;
+
+  private final Optional<String> logFileNamePrefix;
+
+  private final int linesWrittenBeforeFlush;
+
+  private final ScheduledExecutorService logCopyExecutor;
+
+  // A set of current source log files being copied
+  private final Set<Path> srcLogFiles = Sets.newHashSet();
+
+  // A map from source log file paths to the most recent positions the copier have copied to
+  private final Map<Path, Long> currentPosMap = Maps.newHashMap();
```

Comment:
Actually there's some `check-then-act` compound operations that cannot be made atomic even if using `ConcurrentMap`. Given that synchronization is needed anyway for such compound operations, using `ConcurrentMap` doesn't offer any benefit.